packages/react-components/react-motion-preview/src/hooks/useMotion.ts

```diff
@@ -0,0 +1,263 @@
+  const [setDelayedAnimationFrame, cancelDelayedAnimationFrame] = useAnimationFrame();
+  const skipAnimationOnFirstRender = React.useRef(!animateOnFirstMount);
+
+  const processAnimation = React.useCallback(
+    (callback: () => void) => {
+      const targetElement = currentElement;
+
+      if (!targetElement) {
+        return;
+      }
+
+      clearAnimationTimeout();
+      cancelProcessingAnimationFrame();
+      setProcessingAnimationFrame(() => {
+        const duration = getMotionDuration(targetElement);
+
+        if (duration === 0) {
+          callback();
+          return;
+        }
+
+        /**
+         * Use CSS transition duration + 1ms to ensure the animation has finished on both enter and exit states.
+         * This is an alternative to using the `transitionend` event which can be unreliable as it fires multiple times
+         * if the transition has multiple properties.
+         */
+        setAnimationTimeout(() => callback(), duration + 1);
+      });
+
+      return () => {
+        clearAnimationTimeout();
+        cancelProcessingAnimationFrame();
+      };
+    },
+    [
+      cancelProcessingAnimationFrame,
+      clearAnimationTimeout,
+      currentElement,
+      setAnimationTimeout,
+      setProcessingAnimationFrame,
+    ],
+  );
+
+  const ref: React.RefCallback<HTMLElementWithStyledMap<Element>> = React.useCallback(node => {
+    if (!node) {
+      return;
+    }
+
+    setCurrentElement(node);
+  }, []);
+
+  React.useEffect(() => {
+    if (presence) {
+      unstable_batchedUpdates(() => {
+        setType('entering');
+        setActive(skipAnimationOnFirstRender.current ? true : false);
+      });
+    }
+  }, [presence]);
+
+  React.useEffect(() => {
+    const skipAnimation = skipAnimationOnFirstRender.current;
+    const onUnmount = () => {
+      cancelActiveAnimationFrame();
+      cancelDelayedAnimationFrame();
+    };
+
+    setActiveAnimationFrame(() => {
+      unstable_batchedUpdates(() => {
+        setActive(presence);
+        setType(() => {
+          if (skipAnimation) {
+            return presence ? 'idle' : 'unmounted';
+          }
+
+          return presence ? 'entering' : 'exiting';
+        });
+      });
+    });
+
+    if (skipAnimation) {
+      return onUnmount;
+    }
+
+    processAnimation(() => {
+      setType(presence ? 'entered' : 'exited');
+      setDelayedAnimationFrame(() => setType(presence ? 'idle' : 'unmounted'));
+    });
+
+    return onUnmount;
+  }, [
+    cancelActiveAnimationFrame,
+    cancelDelayedAnimationFrame,
+    presence,
+    processAnimation,
+    setActiveAnimationFrame,
+    setDelayedAnimationFrame,
+  ]);
+
+  React.useEffect(() => {
+    skipAnimationOnFirstRender.current = false;
+  }, []);
+
+  return React.useMemo(() => {
+    const canRender = () => type !== 'unmounted';
+    const isActive = () => active;
+
+    return {
+      ref,
+      type,
+      canRender,
+      isActive,
+      hasInternalMotion: true,
```

Comment:
??

Reply:
This was a leftover of an experiment I was making. I reverted the change.